source/common/http/filter_manager.cc

```diff
@@ -183,6 +183,12 @@ bool ActiveStreamFilterBase::commonHandleAfterDataCallback(FilterDataStatus stat
       buffer_was_streaming = status == FilterDataStatus::StopIterationAndWatermark;
       commonHandleBufferData(provided_data);
     } else if (complete() && !hasTrailers() && !bufferedData()) {
+      if (parent_.state_.destroyed_) {
+        // If the stream is destroyed, no need to handle the data buffer or trailers.
+        // This can occur if the filter calls sendLocalReply.
+        return false;
+      }
```

Comment:
I would really prefer this change in a dedicated PR, but if it stays here please have a dedicated core test for this change.

Reply:
Re-opened #15075 as the dedicated PR. Integration test in that PR fails without this short-circuit logic.